### PR TITLE
Dynamically build efficient command to encode AV file

### PR DESCRIPTION
**Pull request by yenaing-oo**

### CODE REVIEW

**Comment by yenaing-oo** [comment]
The error logs can be removed once everything is confirmed to be working as expected.

File: lib/model/QubitDigitalObject.php [hunk 1]
```diff
@@ -2757,7 +2757,36 @@ public static function convertVideoToMp4($originalPath, $newPath, $width = null,
...
-        $command = 'ffmpeg -y -i '.$originalPath.' -ar 44100 -c:v libx264 -pix_fmt yuv420p -c:a aac -movflags +faststart '.$newPath.' 2>&1';
+        // Read stream information
+        $format = self::readStreamInformation($originalPath);
+
+        error_log('Stream information: ' . json_encode($format, JSON_PRETTY_PRINT));
+
+        $videoCodec = isset($format['video']['codec_name']) ? $format['video']['codec_name'] : null;
+        $pixelFormat = isset($format['video']['pix_fmt']) ? $format['video']['pix_fmt'] : null;
+        $audioCodec = isset($format['audio']['codec_name']) ? $format['audio']['codec_name'] : null;
+        $sampleRate = isset($format['audio']['sample_rate']) ? $format['audio']['sample_rate'] : null;
```
Comment (danloveg) [inline lib/model/QubitDigitalObject.php:2809]: I would suggest using these variables to make a boolean argument for re-encoding the video or re-encoding the audio.

Another thing that would require a re-encode is fixing the file type. For example, you could have a file with the proper video and audio encoding, but the file is a `.mkv`, not a `.mp4`. I would recommend figuring out if `ffprobe` can output the file extension and use its output. Here I've used the `pathinfo` function which could be used if FFprobe doesn't output the file extension.

```suggestion
        $videoCodec = isset($format['video']['codec_name']) ? $format['video']['codec_name'] : null;
        $pixelFormat = isset($format['video']['pix_fmt']) ? $format['video']['pix_fmt'] : null;
        $audioCodec = isset($format['audio']['codec_name']) ? $format['audio']['codec_name'] : null;
        $sampleRate = isset($format['audio']['sample_rate']) ? $format['audio']['sample_rate'] : null;
        
        $reencodeVideo = ($videoCodec !== 'h264' || $pixelFormat !== 'yuv420p');
        $reencodeAudio = ($audioCoded !== 'aac' || $sampleRate !== 44100);
        $fixContainer = strtolower(pathinfo($originalPath, PATHINFO_EXTENSION)) !== 'mp4';
```

File: lib/model/QubitDigitalObject.php [hunk 2]
```diff
@@ -2757,7 +2757,36 @@ public static function convertVideoToMp4($originalPath, $newPath, $width = null,
...
+
+        error_log('Video codec: ' . $videoCodec);
+        error_log('Pixel format: ' . $pixelFormat);
+        error_log('Audio codec: ' . $audioCodec);
+        error_log('Sample rate'. $sampleRate);
+
+        // Determine the appropriate ffmpeg command
+        if ($videoCodec === 'h264' && $pixelFormat === 'yuv420p' && $audioCodec === 'aac' && $sampleRate === 44100) {
+            $command = 'ffmpeg -y -i ' . escapeshellarg($originalPath) . ' -c:v copy -c:a copy ' . escapeshellarg($newPath) . ' 2>&1';
```
Comment (danloveg) [inline lib/model/QubitDigitalObject.php]: You can use [`copy()`](https://www.php.net/manual/en/function.copy.php) instead of calling FFmpeg here.

File: lib/model/QubitDigitalObject.php [hunk 3]
```diff
@@ -2757,7 +2757,36 @@ public static function convertVideoToMp4($originalPath, $newPath, $width = null,
...
+            error_log('No encoding needed for video file: ' . $command);
+        } elseif ($videoCodec === 'h264' && $pixelFormat === 'yuv420p') {
+            $command = 'ffmpeg -y -i ' . escapeshellarg($originalPath) . ' -c:v copy -c:a aac -ar 44100 ' . escapeshellarg($newPath) . ' 2>&1';
+            error_log('Audio encoding needed for video file: ' . $command);
+        } elseif ($audioCodec === 'aac' && $sampleRate === 44100) {
+            $command = 'ffmpeg -y -i ' . escapeshellarg($originalPath) . ' -ar 44100 -c:v libx264 -pix_fmt yuv420p -c:a copy ' . escapeshellarg($newPath) . ' 2>&1';
```
Comment (danloveg) [inline lib/model/QubitDigitalObject.php]: I would recommend looking into whether `-movflags +faststart` is required here. I'm not certain what that option does, but it *was* included in the original FFmpeg command.

**Review by danloveg — CHANGES_REQUESTED** [review]
This is a good start! Once we've got the logic to create the command sorted out I'm going to suggest testing this somehow. I don't know what sort of mocking capabilities the AtoM test suite has, but ideally if we could mock `self::readStreamInformation` and mock `exec()` we can see if the proper command is being executed for all the edge cases.

If that's not possible, we could construct a set of test videos and ingest them into our local AtoM instances. We could then run ffprobe against all the generated derivatives and see if their formats match all the required codecs and parameters. The test set of videos could be video without audio, incorrectly encoded video with properly encoded audio, etc.

File: lib/model/QubitDigitalObject.php (hunk 3 above)
Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php]: That option makes sure the "moov atom" (sort of like a table of contents for the video file) is placed at the start of the video file to allow playback of the start of the video while the rest gets downloaded. If it is placed at the end of the file, the whole file has to be downloaded before playback can begin. The "moov atom" is specific to MP4 files only. Having "faststart" is recommended especially for playing videos on the web.

I added a method to check whether the "moov atom" is at the start of the file or not.

Addressed by: https://github.com/danloveg/atom/pull/1/commits/1133f0520b93e759fd60d9a8d31919b311709b47

Comment (danloveg) [reply lib/model/QubitDigitalObject.php]: Great, that makes sense since these are videos that are played from the web.

File: lib/model/QubitDigitalObject.php (hunk 1 above)
Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php:2809]: > I would suggest using these variables to make a boolean argument for re-encoding the video or re-encoding the audio.

I agree, I've addressed that here: https://github.com/danloveg/atom/pull/1/commits/1133f0520b93e759fd60d9a8d31919b311709b47

> Another thing that would require a re-encode is fixing the file type. For example, you could have a file with the proper video and audio encoding, but the file is a `.mkv`, not a `.mp4`. I would recommend figuring out if `ffprobe` can output the file extension and use its output. Here I've used the `pathinfo` function which could be used if FFprobe doesn't output the file extension.

That's a good idea. Some files might have a non-MP4 file extension with a MP4 container, or an MP4 extension with the wrong container, but in both cases have the right video and audio encoding. I've made a method that checks whether both criterias are met (right extension and right container) here: https://github.com/danloveg/atom/pull/1/commits/1133f0520b93e759fd60d9a8d31919b311709b47

Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php:2809]: I renamed that method to something more fitting here: 
https://github.com/danloveg/atom/pull/1/commits/6156ab4ac12641f9b5a4520c21bffb124a0f70db

File: lib/model/QubitDigitalObject.php (hunk 2 above)
Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php]: Addressed by:
https://github.com/danloveg/atom/pull/1/commits/1d5b9233328b0ba253f51e595419e068b7f3e90f

**Comment by yenaing-oo** [comment]
> If that's not possible, we could construct a set of test videos and ingest them into our local AtoM instances. We could then run ffprobe against all the generated derivatives and see if their formats match all the required codecs and parameters. The test set of videos could be video without audio, incorrectly encoded video with properly encoded audio, etc.

I am currently generating a set of test video files for each test case I can think of using `ffmpeg`.

> Once we've got the logic to create the command sorted out I'm going to suggest testing this somehow. I don't know what sort of mocking capabilities the AtoM test suite has, but ideally if we could mock `self::readStreamInformation` and mock `exec()` we can see if the proper command is being executed for all the edge cases.

I'll work on this once I'm satisfied with the behaviour of the code.

**Comment by danloveg** [comment]
> > If that's not possible, we could construct a set of test videos and ingest them into our local AtoM instances. We could then run ffprobe against all the generated derivatives and see if their formats match all the required codecs and parameters. The test set of videos could be video without audio, incorrectly encoded video with properly encoded audio, etc.
> 
> I am currently generating a set of test video files for each test case I can think of using `ffmpeg`.

If you need help uploading video files to your test instance, let me know. It won't let you upload files beyond a certain size from the frontend, but you can use [this tool from the backend](https://www.accesstomemory.org/en/docs/2.8/admin-manual/maintenance/cli-import-export/#load-digital-objects-via-the-command-line).

**Comment by danloveg** [comment]
I updated the `readStreamInformation` function to include the format information. You shouldn't need to call `ffprobe` again in the `isCorrectMP4Format` function.

**Comment by yenaing-oo** [comment]
> I updated the `readStreamInformation` function to include the format information. You shouldn't need to call `ffprobe` again in the `isCorrectMP4Format` function.

That's great, I'll adapt my code to make use of `format_name` from `readStreamInformation` instead.

**Comment by danloveg** [comment]
I had an idea. Since it looks like you're using FFmpeg to copy the file even when no re-encoding is required (which is a good change, since it'll fix any issues with the container), we could probably *always* include `-movflags +faststart`. It sounds like FFmpeg will skip that operation if it doesn't need to be done, which means that it does the checking for us already.

**Comment by yenaing-oo** [comment]
> It sounds like FFmpeg will skip that operation if it doesn't need to be done, which means that it does the checking for us already.

I couldn't find this being mentioned anywhere explicitly. I read that using `-movflags +faststart` can add a few seconds if the file is large:
https://superuser.com/questions/856025/any-downsides-to-always-using-the-movflags-faststart-parameter

If `ffmpeg` does not perform a check for us to see if fast-starting is already applied, we would be doing it again unnecessarily.

I restructured my code again to account for the case where no reformatting, reencoding, or fast-starting is needed, in which case a simple PHP copy would suffice: https://github.com/danloveg/atom/pull/1/commits/f90359321d1abf24d56a305149ab098c7028a66a

**Comment by danloveg** [comment]
> > It sounds like FFmpeg will skip that operation if it doesn't need to be done, which means that it does the checking for us already.
> 
> I couldn't find this being mentioned anywhere explicitly. I read that using `-movflags +faststart` can add a few seconds if the file is large: https://superuser.com/questions/856025/any-downsides-to-always-using-the-movflags-faststart-parameter
> 
> If `ffmpeg` does not perform a check for us to see if fast-starting is already applied, we would be doing it again unnecessarily.
> 
> I restructured my code again to account for the case where no reformatting, reencoding, or fast-starting is needed, in which case a simple PHP copy would suffice: [f903593](https://github.com/danloveg/atom/commit/f90359321d1abf24d56a305149ab098c7028a66a)

That makes sense, I see where you're coming from. Let's keep checking for that flag at the start of the file then, just thought we could maybe simplify all the checks.

**Comment by yenaing-oo** [comment]
@danloveg 

I'm having trouble mocking built-in PHP functions such as `exec()`, `copy()` and `chmod()`. I think PHP does not allow mocking of built-in functions. One workaround that I've seen is to use a wrapper function, for e.g., `doCopy()` which calls `copy()`, instead of calling `copy` directly, and mock `doCopy()` instead.

I was wondering if you knew a better solution to deal with this. I tried a potential solution using namespaces, but I couldn't get it to work: https://stackoverflow.com/a/56386772

Here's what I have so far: https://github.com/danloveg/atom/pull/1/commits/fdb2c41ab9115fe603f6a98f19f7e3004fde6342

EDIT: In case you don't have the command to run units tests handy: `docker compose exec atom composer test`

**Comment by danloveg** [comment]
> @danloveg
> 
> I'm having trouble mocking built-in PHP functions such as `exec()`, `copy()` and `chmod()`. I think PHP does not allow mocking of built-in functions. One workaround that I've seen is to use a wrapper function, for e.g., `doCopy()` which calls `copy()`, instead of calling `copy` directly, and mock `doCopy()` instead.
> 
> I was wondering if you knew a better solution to deal with this. I tried a potential solution using namespaces, but I couldn't get it to work: https://stackoverflow.com/a/56386772
> 
> Here's what I have so far: [fdb2c41](https://github.com/danloveg/atom/commit/fdb2c41ab9115fe603f6a98f19f7e3004fde6342)
> 
> EDIT: In case you don't have the command to run units tests handy: `docker compose exec atom composer test`

That's fine, I thought that might be an issue.

I would prefer not to include any more test files as much as we can avoid it. I would also prefer we not include any calls to external programs in the unit tests (see the last paragraph in this comment). Instead of testing that `ffprobe` and `ffmpeg` output the right thing, we could try this:

- Mock `readStreamInformation`
- Mock `isFastStarted`
- Create a new function, `public static getConvertToMp4Command(string $filePath): string`
- Instead of testing `convertVideoToMp4`, test this new function.

For testing whether `readStreamInformation` or `convertVideoToMp4` works correctly, you may want to create a functional test instead. But I would say we don't need to do this unless Artefactual asks us to write a test like that to verify the change. Including unit tests is already going above and beyond what a typical pull request to atom has in it.

Edit: "Test files" being new files in `fixtures/` - just read this back and it might sound like I meant no new `.php` files.

**Comment by yenaing-oo** [comment]
@danloveg 

I've run into another roadblock - it appears that there is no way to mock static methods using PHPUnit.
https://stackoverflow.com/a/2357141

I don't think any alternative was provided when they deprecated the once working `staticExpects()` method:
https://github.com/sebastianbergmann/phpunit-documentation/issues/77#issuecomment-25429003

The way around this is to make all the methods non-static (`getConvertToMp4Command`, `readStreamInformation`, `isFastStarted`), but this would involve changing the method signature of `convertVideoToMp4` as well to make it non-static. I don't think Artefactual would allow that.

**Comment by danloveg** [comment]
> @danloveg
> 
> I've run into another roadblock - it appears that there is no way to mock static methods using PHPUnit. https://stackoverflow.com/a/2357141
> 
> I don't think any alternative was provided when they deprecated the once working `staticExpects()` method: [sebastianbergmann/phpunit-documentation#77 (comment)](https://github.com/sebastianbergmann/phpunit-documentation/issues/77#issuecomment-25429003)
> 
> The way around this is to make all the methods non-static (`getConvertToMp4Command`, `readStreamInformation`, `isFastStarted`), but this would involve changing the method signature of `convertVideoToMp4` as well. I don't think Artefactual would allow that.

Lovely! Okay, thanks for all that detective work. How about this, we include your package of test files in the pull request, and say that those are the files we tested this on and it is working on at least those files. It needs further testing, but we had trouble writing unit tests due to being unable to mock static methods.

**Comment by yenaing-oo** [comment]
> Lovely! Okay, thanks for all that detective work. How about this, we include your package of test files in the pull request, and say that those are the files we tested this on and it is working on at least those files. It needs further testing, but we had trouble writing unit tests due to being unable to mock static methods.

Sounds like the best choice for now. How should we include those file? Perhaps using a OneDrive link?

**Comment by danloveg** [comment]
> > Lovely! Okay, thanks for all that detective work. How about this, we include your package of test files in the pull request, and say that those are the files we tested this on and it is working on at least those files. It needs further testing, but we had trouble writing unit tests due to being unable to mock static methods.
> 
> Sounds like the best choice for now. How should we include those file? Perhaps using a OneDrive link?

I can just upload them like this in the pull request comment:

[test_vid_files.zip](https://github.com/user-attachments/files/18759217/test_vid_files.zip)

**Comment by yenaing-oo** [comment]
> > > Lovely! Okay, thanks for all that detective work. How about this, we include your package of test files in the pull request, and say that those are the files we tested this on and it is working on at least those files. It needs further testing, but we had trouble writing unit tests due to being unable to mock static methods.
> > 
> > 
> > Sounds like the best choice for now. How should we include those file? Perhaps using a OneDrive link?
> 
> I can just upload them like this in the pull request comment:
> 
> [test_vid_files.zip](https://github.com/user-attachments/files/18759217/test_vid_files.zip)

Ah that's perfect. Ok, I'll revert the last 3 commits involving tests.

**Comment by danloveg** [comment]
> > > > Lovely! Okay, thanks for all that detective work. How about this, we include your package of test files in the pull request, and say that those are the files we tested this on and it is working on at least those files. It needs further testing, but we had trouble writing unit tests due to being unable to mock static methods.
> > > 
> > > 
> > > Sounds like the best choice for now. How should we include those file? Perhaps using a OneDrive link?
> > 
> > 
> > I can just upload them like this in the pull request comment:
> > [test_vid_files.zip](https://github.com/user-attachments/files/18759217/test_vid_files.zip)
> 
> Ah that's perfect. Ok, I'll revert the last 3 commits involving tests.

Sounds good, feel free to request a review after that, I have a couple small comments just to clean the changes up a bit

File: lib/model/QubitDigitalObject.php [hunk 4]
```diff
@@ -2743,6 +2747,38 @@ public static function readStreamInformation(string $filePath): array
...
+     * Check if the video file's moov atom is at the front of the file.
+     *
+     * @param string $filePath path to the video file
+     *
+     * @return bool true if moov atom is at the front, false otherwise or if moov atom does not exist, which means the
+     * file is not a valid MP4 file
+     */
+    public static function isFastStarted($filePath)
+    {
+        $command = "ffmpeg -v trace -i " . escapeshellarg($filePath) . " 2>&1 | grep -e type:\\'mdat\\' -e type:\\'moov\\'";
```
Comment (danloveg) [inline lib/model/QubitDigitalObject.php]: We can make guarantees about `ffmpeg` and `ffprobe` being available on the system, but we can't say the same about `grep`. I'd recommend removing the pipe to grep, and processing the raw output using PHP instead.

You'll probably also want to track when both the moov atom and mdat atoms are found, and break out of the `foreach` loop when both are seen.

File: lib/model/QubitDigitalObject.php [hunk 5]
```diff
@@ -2764,8 +2800,65 @@ public static function convertVideoToMp4($originalPath, $newPath, $width = null,
...
+            $command = 'ffmpeg -y -i ' . escapeshellarg($originalPath) . ' -c:v copy -c:a aac -ar 44100 ' . escapeshellarg($newPath);
+            error_log('Audio encoding needed for video file: ' . $command);
+        } elseif ($reencodeVideo) {
+            $command = 'ffmpeg -y -i ' . escapeshellarg($originalPath) . ' -c:v libx264 -pix_fmt yuv420p -c:a copy ' . escapeshellarg($newPath);
+            error_log('Video encoding needed for video file: ' . $command);
+        } elseif ($reformatFile) {
+            // All the above commands also reformat the file
+            error_log('Reformatting needed for video file');
+            $command = $default_command;
+        }
```
Comment (danloveg) [inline lib/model/QubitDigitalObject.php]: How do you feel about constructing the options like this instead?

```suggestion
        $codecOptions = '';

        // Determine the appropriate ffmpeg command
        if ($reencodeVideo && $reencodeAudio) {
            $codecOptions = '-c:v libx264 -pix_fmt yuv420p -c:a aac -ar 44100';
        } elseif ($reencodeAudio) {
            $codecOptions = '-c:v copy -c:a aac -ar 44100';
        } elseif ($reencodeVideo) {
            $codecOptions = '-c:v libx264 -pix_fmt yuv420p -c:a copy';
        } elseif ($reformatFile) {
            $codecOptions = '-c copy';
        }
```

And later, if you need faststart, you can do `$codecOptions .= ' -movflags +faststart';`. If there were some codec options, then you can build the command like this:

```php
if ($codecOptions) {
    $command = sprintf(
        'ffmpeg -y -i %s %s %s 2>&1',
        escapeshellarg($originalPath),
        $codecOptions,
        escapeshellarg($newPath),
    );
    // Then exec() it
}
else {
    // copy() it
}
```

I think there's a bit of an opportunity to make this a bit cleaner. You don't have to take these changes verbatim but I think they would help make this a bit more readable.

**Review by danloveg — CHANGES_REQUESTED** [review]

File: lib/model/QubitDigitalObject.php (hunk 4 above)
Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php]: It took me a while to figure out that the status code returned by executing `ffmpeg -v trace -i " . escapeshellarg($filePath) . " 2>&1` is `1` (error). 

The way this command is used, an output file is intentionally left out so that the input file is not processed all the way - `ffmpeg` simply reads the file and logs its steps in detail before quitting once it realizes there is no output file.

The change I made should also exit the loop early once the first `mdat` atom is found, by which point it should be able to tell whether a `moov` atom had already been found or not.

Addressed by: https://github.com/danloveg/atom/pull/1/commits/ee1874343661c41395f0e101570b27b5de5b4633

File: lib/model/QubitDigitalObject.php (hunk 5 above)
Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php]: I think the code looks much neater this way.

I've applied your suggested changes here: https://github.com/danloveg/atom/pull/1/commits/3610178987d691cb6d304a9edbb2e8d72ec4e7ef

Let me know if you see any opportunity to clean it up further.

File: lib/model/QubitDigitalObject.php (hunk 4 above)
Comment (danloveg) [reply lib/model/QubitDigitalObject.php]: Would it work to swap ffmpeg for ffprobe in that command? (Minus the `-i` option). I don't know if ffprobe would create the same output but at least it'd return a 0 status.

Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php]: You're right, `ffprobe` works perfectly in place of `ffmpeg` for this purpose, without returning an error. I've applied that change here:
https://github.com/danloveg/atom/pull/1/commits/56fb03e5823600212287c1e626f23af4f452119f and https://github.com/danloveg/atom/pull/1/commits/8bef4f13c72f3b044c3532365b5115381a3e6e54

File: lib/model/QubitDigitalObject.php (hunk 5 above)
Comment (danloveg) [reply lib/model/QubitDigitalObject.php]: Looks good to me!

File: lib/model/QubitDigitalObject.php (hunk 4 above)
Comment (danloveg) [reply lib/model/QubitDigitalObject.php]: Okay great, I wasn't certain whether that'd be the case

File: lib/model/QubitDigitalObject.php [hunk 6]
```diff
@@ -2764,8 +2800,70 @@ public static function convertVideoToMp4($originalPath, $newPath, $width = null,
...
+        $pixelFormat = isset($format['video']['pix_fmt']) ? $format['video']['pix_fmt'] : null;
+        $audioCodec = isset($format['audio']['codec_name']) ? $format['audio']['codec_name'] : null;
+        $sampleRate = isset($format['audio']['sample_rate']) ? $format['audio']['sample_rate'] : null;
+        $format_name = isset($format['format_name']) ? $format['format_name'] : null;
+
+        $reencodeVideo = ($videoCodec !== 'h264' || $pixelFormat !== 'yuv420p');
+        $reencodeAudio = ($audioCodec !== 'aac' || $sampleRate !== 44100);
+        $reformatFile = strpos($format_name, 'mp4') === false || strtolower(pathinfo($originalPath, PATHINFO_EXTENSION)) !== 'mp4';
+
+        $needFastStart = !self::isFastStarted($originalPath);
```
Comment (danloveg) [inline lib/model/QubitDigitalObject.php]: I think you can short-circuit this check here to skip running the `ffprobe` command again in certain situations:

```suggestion
        $needFastStart = $reencodeVideo || !self::isFastStarted($originalPath);
```

Since if you're re-encoding the video, you'll want to faststart it by default.

Comment (yenaing-oo) [reply lib/model/QubitDigitalObject.php]: I believe it is possible for some MP4 files to be faststarted already, but do not meet the video encoding criteria. For e.g., here is a MP4 file that is already faststarted and has `hevc` video encoding:


https://github.com/user-attachments/assets/cdf7a182-bb15-4af3-b86f-fff583447ab4

```bash
╰─ ffprobe -v quiet -show_entries format=format_name -show_streams -print_format json invalid_vid_codec_faststarted.mp4 2>&1
{
    "streams": [
        {
            "index": 0,
            "codec_name": "hevc",
            "codec_long_name": "H.265 / HEVC (High Efficiency Video Coding)",
            "profile": "Main",
            "codec_type": "video",
            "codec_tag_string": "hev1",
            "codec_tag": "0x31766568",
            "width": 480,
            "height": 270,
            "coded_width": 480,
            "coded_height": 272,
            "closed_captions": 0,
            "film_grain": 0,
            "has_b_frames": 2,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "16:9",
            "pix_fmt": "yuv420p",
            "level": 63,
            "color_range": "tv",
            "chroma_location": "left",
            "field_order": "progressive",
            "refs": 1,
            "view_ids_available": "",
            "view_pos_available": "",
            "id": "0x1",
            "r_frame_rate": "30/1",
            "avg_frame_rate": "3375/113",
            "time_base": "1/15360",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 461824,
            "duration": "30.066667",
            "bit_rate": "27420",
            "nb_frames": "900",
            "extradata_size": 2466,
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "non_diegetic": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0,
                "multilayer": 0
            },
            "tags": {
                "language": "und",
                "handler_name": "VideoHandler",
                "vendor_id": "[0][0][0][0]",
                "encoder": "Lavc61.19.100 libx265"
            }
        },
        {
            "index": 1,
            "codec_name": "aac",
            "codec_long_name": "AAC (Advanced Audio Coding)",
            "profile": "LC",
            "codec_type": "audio",
            "codec_tag_string": "mp4a",
            "codec_tag": "0x6134706d",
            "sample_fmt": "fltp",
            "sample_rate": "44100",
            "channels": 2,
            "channel_layout": "stereo",
            "bits_per_sample": 0,
            "initial_padding": 0,
            "id": "0x2",
            "r_frame_rate": "0/0",
            "avg_frame_rate": "0/0",
            "time_base": "1/44100",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 1348578,
            "duration": "30.580000",
            "bit_rate": "2093",
            "nb_frames": "1318",
            "extradata_size": 5,
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "non_diegetic": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0,
                "multilayer": 0
            },
            "tags": {
                "language": "und",
                "handler_name": "SoundHandler",
                "vendor_id": "[0][0][0][0]"
            }
        }
    ],
    "format": {
        "format_name": "mov,mp4,m4a,3gp,3g2,mj2"
    }
}

╭─    ~/Desktop/vid_test/valid ··························· ✔  09:03:12 
╰─ ffprobe -v trace invalid_vid_codec_faststarted.mp4 2>&1 | grep -E "type:'(mdat|moov)'"
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x11fe04e50] type:'moov' parent:'root' sz: 37421 36 148532
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x11fe04e50] type:'mdat' parent:'root' sz: 111075 37465 148532

```

Therefore, requiring video reencoding does not always necessitate faststarting.

Comment (danloveg) [reply lib/model/QubitDigitalObject.php]: That makes sense, disregard my comment then! Still just looking at small ways to make efficiencies, looks like we're approaching a state where it's as efficient as possible. I would love if there were a way to output the moov/mdat atom information when I run FFprobe the first time but I don't think that's possible.

**Comment by yenaing-oo** [comment]
I've removed the error logs here: https://github.com/danloveg/atom/pull/1/commits/590d6354352d3009ce7d46b02bfe4c9f52bb6014

**Comment by danloveg** [comment]
> I've removed the error logs here: [590d635](https://github.com/danloveg/atom/commit/590d6354352d3009ce7d46b02bfe4c9f52bb6014)

Thanks, I'll merge this in, close this PR, and then you can tackle the audio part.
